src/utils.ts

```diff
@@ -1198,4 +1352,786 @@ export class RubberBand {
+  }
+
+  private marginedBounds(minValue: number, maxValue: number): [number, number] {
+    if (this.isDiscrete) {return [minValue - 1, maxValue + 1]};
+    return [minValue * (1 - Math.sign(minValue) * this.marginRatio),
+            maxValue * (1 + Math.sign(maxValue) * this.marginRatio)];
+  }
+
+  public drawRubberBand(context: CanvasRenderingContext2D) {
+    const realMin = this.relativeToAbsolute(this.rubberBand.minValue);
+    const realMax = this.relativeToAbsolute(this.rubberBand.maxValue);
+    const coord = this.isVertical ? "y" : "x";
+    this.rubberBand.realMin = Math.max(Math.min(realMin, realMax), this.origin[coord]);
+    this.rubberBand.realMax = Math.min(Math.max(realMin, realMax), this.end[coord]);
+    this.rubberBand.realMin = Math.min(this.rubberBand.realMin, this.rubberBand.realMax);
+    this.rubberBand.realMax = Math.max(this.rubberBand.realMin, this.rubberBand.realMax);
+    this.rubberBand.draw(this.isVertical ? this.origin.x : this.origin.y, context, this.rubberColor, 'hsl(203, 0%, 100%, 0.5)', 0.1, 1.);
+  }
+
+  public mouseMove(mouseDown: Vertex, mouseCoords: Vertex) {
+    let downValue = this.absoluteToRelative(this.isVertical ? mouseDown.y : mouseDown.x);
+    let currentValue = this.absoluteToRelative(this.isVertical ? mouseCoords.y : mouseCoords.x);
+    if (!this.rubberBand.isClicked) {
+      this.rubberBand.minValue = Math.min(downValue, currentValue);
+      this.rubberBand.maxValue = Math.max(downValue, currentValue);
+    } else {this.rubberBand.mouseMove(downValue, currentValue)}
+    return true
+  }
+
+  public mouseDown(mouseDown: Vertex) {
+    let isReset = false;
+    const mouseUniCoord = this.isVertical ? mouseDown.y : mouseDown.x;
+    if (!this.isInRubberBand(this.absoluteToRelative(mouseUniCoord))) {
+      this.rubberBand.reset();
+      isReset = true;
+    } else {this.rubberBand.mouseDown(mouseUniCoord)}
+    return isReset
+  }
+
+  public mouseUp() {this.rubberBand.mouseUp()}
+
+  public isInRubberBand(value: number): boolean {
+    return (value >= this.rubberBand.minValue && value <= this.rubberBand.maxValue) ? true : false
+  }
+
+  public numericLabels(): string[] {
+    this.updateTickPrecision();
+    let numericLabels = []
+    this.ticks.forEach(tick => numericLabels.push(tick.toPrecision(this.tickPrecision)));
+    return numericLabels
+  }
+
+  public saveLoc(): void {
+    this._previousMin = this.minValue;
+    this._previousMax = this.maxValue;
+  }
+
+  public stringsToValues(vector: any[]): number[] {
+    if (this.isDiscrete) {
+      let numericVector = [];
+      vector.forEach((value) => numericVector.push(this.labels.indexOf(value)));
+      return numericVector
+    }
+    return vector
+  }
+
+  private textAlignments(): [string, string] {
+    const forVertical = ['right', 'left'][Math.sign(1 - Math.sign(this.transformMatrix.e))];
+    const forHorizontal = ['hanging', 'alphabetic'][Math.sign(1 - Math.sign(this.transformMatrix.f))]
```

Comment:
Same comment as before, plus, this could be set as a function, as this is duplicate

Reply:
Done this time